File: packages/bdb/src/tests/database.test.ts
import { MockStorage } from 'storage';
import { BsonDatabase } from '../lib/database';
import { RandomUuidGenerator, TimestampProvider } from 'utils';

//
// Valid UUID v4 strings for use as record IDs in tests.
//
const ID1 = '11111111-1111-4111-a111-111111111111';
const ID2 = '22222222-2222-4222-a222-222222222222';
const ID3 = '33333333-3333-4333-a333-333333333333';

describe('BsonDatabase', () => {
    let storage: MockStorage;
    let database: BsonDatabase;

    beforeEach(() => {
        storage = new MockStorage();
        database = new BsonDatabase(storage, "", new RandomUuidGenerator(), new TimestampProvider());
    });

    //
    // collection()
    //

    test('should create a new collection', () => {
        const collection = database.collection('users');
        expect(collection).toBeDefined();
    });

    test('should return the same collection instance for the same name', () => {
        const collection1 = database.collection('users');
        const collection2 = database.collection('users');
        expect(collection1).toBe(collection2);
    });

    test('should return different collection instances for different names', () => {
        const collection1 = database.collection('users');
        const collection2 = database.collection('products');
        expect(collection1).not.toBe(collection2);
    });

    //
    // collections()
    //

    test('should list collections created in memory', async () => {
        database.collection('users');
        database.collection('products');
        database.collection('orders');

        const collections = await database.collections();

        expect(collections).toContain('users');
        expect(collections).toContain('products');
        expect(collections).toContain('orders');
    });

    test('should list collections from storage', async () => {
        await storage.write('collections/metadata/shards/0', undefined, Buffer.from('test'));

        const collections = await database.collections();

        expect(collections).toContain('metadata');
    });

    test('should merge in-memory and storage collections without duplicates', async () => {
        database.collection('users');
        await storage.write('collections/users/shards/0', undefined, Buffer.from('test'));
        await storage.write('collections/photos/shards/0', undefined, Buffer.from('test'));

        const collections = await database.collections();

        const userEntries = collections.filter(name => name === 'users');
        expect(userEntries).toHaveLength(1);
        expect(collections).toContain('photos');
    });

    test('should return empty array when no collections exist', async () => {
        const collections = await database.collections();
        expect(collections).toHaveLength(0);
    });

    //
    // commit()
    //

    test('commit should be a no-op when not dirty', async () => {
        // Should not throw and should complete quickly
        await database.commit();
    });

    test('commit should allow flush after completing', async () => {
        const users = database.collection<{ _id: string; name: string }>('users');
        await users.insertOne({ _id: ID1, name: 'Alice' });

        await database.commit();

        // After commit, flush should succeed (dirty flag cleared)
        expect(() => database.flush()).not.toThrow();
    });

    test('commit clears dirty flag so a second commit is a no-op', async () => {
        const users = database.collection<{ _id: string; name: string }>('users');
        await users.insertOne({ _id: ID1, name: 'Alice' });

        await database.commit();

        // Second commit should not throw and should be a no-op
        await database.commit();
    });

    test('commit processes multiple collections', async () => {
        const users = database.collection<{ _id: string; name: string }>('users');
        const products = database.collection<{ _id: string; title: string }>('products');

        await users.insertOne({ _id: ID1, name: 'Alice' });
        await products.insertOne({ _id: ID2, title: 'Widget' });

        // Should not throw when committing multiple collections
        await database.commit();
        expect(() => database.flush()).not.toThrow();
    });

    //
    // flush()
    //

    test('flush should throw when database is dirty', async () => {
        const users = database.collection<{ _id: string; name: string }>('users');
        await users.insertOne({ _id: ID1, name: 'Alice' });

        expect(() => database.flush()).toThrow();
    });

    test('flush should succeed when database is not dirty', () => {
        // Never modified — flush should not throw
        expect(() => database.flush()).not.toThrow();
    });

    test('flush should succeed after commit', async () => {
        const users = database.collection<{ _id: string; name: string }>('users');
        await users.insertOne({ _id: ID1, name: 'Alice' });

        await database.commit();
        expect(() => database.flush()).not.toThrow();
    });

    test('flush evicts collection internal state but keeps the collection object cached', async () => {
        const users = database.collection<{ _id: string; name: string }>('users');
        await users.insertOne({ _id: ID1, name: 'Alice' });
        await database.commit();

        database.flush();

        // The same collection instance is returned (collection objects stay cached)
        const users2 = database.collection<{ _id: string; name: string }>('users');
        expect(users2).toBe(users);
    });

    //
    // merkleTree()
    //

    test('merkleTree should return an IMerkleRef', () => {
        const ref = database.merkleTree();
        expect(ref).toBeDefined();
        expect(typeof ref.get).toBe('function');
        expect(typeof ref.upsert).toBe('function');
        expect(typeof ref.remove).toBe('function');
        expect(typeof ref.commit).toBe('function');
        expect(typeof ref.flush).toBe('function');
    });

    test('merkleTree should return the same instance on repeated calls', () => {
        const ref1 = database.merkleTree();
        const ref2 = database.merkleTree();
        expect(ref1).toBe(ref2);
    });

    test('merkleTree get should return undefined for an empty database', async () => {
        const tree = await database.merkleTree().get();
        expect(tree).toBeUndefined();
    });

    test('merkleTree should return a new instance after flush', async () => {
        const ref1 = database.merkleTree();
        // Must commit before flushing (even though nothing is dirty)
        database.flush();
        const ref2 = database.merkleTree();
        expect(ref2).not.toBe(ref1);
    });
});

File: packages-zig/bdb-zig/src/test/database.test.zig
const std = @import("std");
const bdb = @import("bdb-zig");
const utils = @import("utils-zig");
const serialization_zig = @import("serialization-zig");
const helpers = @import("test-helpers.zig");
const MemoryStorage = @import("memory-storage.zig").MemoryStorage;
const BsonDocument = serialization_zig.bson.BsonDocument;
const BsonDatabase = bdb.database.BsonDatabase;
const IInternalRecord = bdb.shard.IInternalRecord;

const io = std.testing.io;

//
// Valid UUID v4 strings for use as record IDs in tests.
//
const ID1 = "11111111-1111-4111-a111-111111111111";

//
// A second record id.
//
const ID2 = "22222222-2222-4222-a222-222222222222";

//
// Generates tree ids.
//
var test_uuid_generator: utils.test_uuid_generator.TestUuidGenerator = .{};

//
// Creates the database under test (TypeScript: the beforeEach block).
//
fn newDatabase(allocator: std.mem.Allocator, storage: *MemoryStorage) !*BsonDatabase {
    return BsonDatabase.init(allocator, storage.asStorage(), "", test_uuid_generator.uuidGenerator(), helpers.timestamp_provider.timestampProvider());
}

//
// Builds a record with one string field.
//
fn makeRecord(allocator: std.mem.Allocator, id: []const u8, key: []const u8, value: []const u8) !IInternalRecord {
    return .{
        ._id = id,
        .fields = try BsonDocument.fromFields(allocator, &.{.{ .key = key, .value = .{ .string = value } }}),
        .metadata = .empty,
    };
}

test "should create a new collection" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var storage = MemoryStorage.init(arena.allocator());
    const database = try newDatabase(arena.allocator(), &storage);
    const collection = try database.collection("users");
    try std.testing.expectEqualStrings("users", collection.name);
}

test "should return the same collection instance for the same name" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var storage = MemoryStorage.init(arena.allocator());
    const database = try newDatabase(arena.allocator(), &storage);
    try std.testing.expectEqual(try database.collection("users"), try database.collection("users"));
}

test "should return different collection instances for different names" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var storage = MemoryStorage.init(arena.allocator());
    const database = try newDatabase(arena.allocator(), &storage);
    try std.testing.expect(try database.collection("users") != try database.collection("products"));
}

test "commit should be a no-op when not dirty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var storage = MemoryStorage.init(arena.allocator());
    const database = try newDatabase(arena.allocator(), &storage);
    try database.commit(io);
    try std.testing.expectEqual(@as(usize, 0), storage.files.count());
}

test "commit clears dirty flag so a second commit is a no-op" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    const database = try newDatabase(allocator, &storage);
    const users = try database.collection("users");
    try users.setInternalRecord(io, try makeRecord(allocator, ID1, "name", "Alice"));
    try std.testing.expect(database.dirty);

    try database.commit(io);
    try std.testing.expect(!database.dirty);
    const databaseTree = storage.getFile("db.dat").?;

    // Second commit should not throw and should be a no-op
    try database.commit(io);
    try std.testing.expectEqual(databaseTree.ptr, storage.getFile("db.dat").?.ptr);
}

test "commit processes multiple collections" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    const database = try newDatabase(allocator, &storage);
    const users = try database.collection("users");
    const products = try database.collection("products");
    try users.setInternalRecord(io, try makeRecord(allocator, ID1, "name", "Alice"));
    try products.setInternalRecord(io, try makeRecord(allocator, ID2, "title", "Widget"));

    try database.commit(io);
    try std.testing.expect(!users.dirty());
    try std.testing.expect(!products.dirty());
    try std.testing.expect(storage.getFile("collections/users/collection.dat") != null);
    try std.testing.expect(storage.getFile("collections/products/collection.dat") != null);
    const databaseTree = (try (try database.merkleTree()).get(io)).?;
    try std.testing.expectEqual(@as(u32, 2), databaseTree.sort.?.leafCount);
}

test "merkleTree should return the same instance on repeated calls" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var storage = MemoryStorage.init(arena.allocator());
    const database = try newDatabase(arena.allocator(), &storage);
    try std.testing.expectEqual(try database.merkleTree(), try database.merkleTree());
}

test "merkleTree get should return undefined for an empty database" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var storage = MemoryStorage.init(arena.allocator());
    const database = try newDatabase(arena.allocator(), &storage);
    try std.testing.expect((try (try database.merkleTree()).get(io)) == null);
}

test "commit removes a collection from the database tree when its last record is deleted" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    const database = try newDatabase(allocator, &storage);
    const users = try database.collection("users");
    try users.setInternalRecord(io, try makeRecord(allocator, ID1, "name", "Alice"));
    try database.commit(io);
    try std.testing.expect(storage.getFile("db.dat") != null);

    _ = try users.deleteOne(io, ID1);
    try database.commit(io);
    try std.testing.expect(storage.getFile("db.dat") == null);
    try std.testing.expect(storage.getFile("collections/users/collection.dat") == null);
}

test "a database committed by Zig has the root hash TypeScript computes for the same records" {
    // The 50-assets test database was written by TypeScript. Re-setting every record into an empty database gives
    // the same collection root hash, because merkle hashes only depend on the record hashes.
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var source = MemoryStorage.init(allocator);
    try source.loadDirectory(io, helpers.TEST_DBS_DIR ++ "/50-assets/.db/bson", ".db/bson");
    var destination = MemoryStorage.init(allocator);
    const sourceDatabase = try BsonDatabase.init(allocator, source.asStorage(), ".db/bson", test_uuid_generator.uuidGenerator(), helpers.timestamp_provider.timestampProvider());
    const destinationDatabase = try BsonDatabase.init(allocator, destination.asStorage(), ".db/bson", test_uuid_generator.uuidGenerator(), helpers.timestamp_provider.timestampProvider());
    const sourceCollection = try sourceDatabase.collection("metadata");
    const destinationCollection = try destinationDatabase.collection("metadata");
    var iterator = sourceCollection.iterateRecords();
    while (try iterator.next(io)) |record| {
        try destinationCollection.setInternalRecord(io, record);

        // Committed after every record: the collection keeps at most 8 shards cached and drops a newly created
        // shard straight away when the other 8 are dirty, which would lose its record (as in TypeScript).
        try destinationDatabase.commit(io);
    }

    const sourceTree = (try bdb.merkle_tree.loadDatabaseMerkleTree(allocator, io, source.asStorage(), ".db/bson")).?;
    const destinationTree = (try bdb.merkle_tree.loadDatabaseMerkleTree(allocator, io, destination.asStorage(), ".db/bson")).?;
    try std.testing.expectEqualSlices(u8, sourceTree.merkle.?.hash, destinationTree.merkle.?.hash);
}
